### ISSUE TITLE
cephfs volume times out during container start, due to recursive selinux relabelling

### ISSUE THREAD

**Issue by anguslees** [rook/rook#2177]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

I have a rook/cephfs volume that used to work fine, but now takes ~10 minutes per container, which hits all manner of timeouts and the pod now fails to start.

I believe (due to the below) the trigger was copying a bunch of existing data into the cephfs volume - it now contains ~60k files.

I'm new to ceph, but I see the mds periodically flooded with thousands of setxattr calls.  This triggers ceph health warnings about operations taking >30s, and occasionally DoSes dockerd (kubelet complains PLEG is not responding and temporarily goes NotReady, causing the manager to churn osd/mon pods).

From digging through code and internet discussions, I've learned that:
- the kubelet will recursively walk a filesystem doing chown/chmod if fsGroup is set.  See https://github.com/kubernetes/kubernetes/blob/43ce5fbf41411babfacd1f5dda9e30fb19691f1e/pkg/volume/flexvolume/mounter.go#L96
- docker will do a recursive selinux relabel if the platform supports selinux (mine does - coreos, and I note the host's ceph mount options include `seclabel`).  I haven't tracked down the specific code yet.

Given that I'm seeing lots of setxattr calls, I believe I'm hitting the latter (selinux relabel).

**Expected behavior:**

cephfs volumes continue to be equally usable regardless of the number of files in the filesystem.

**How to reproduce it (minimal and precise):**

Create a rook/cephfs volume that contains lots of files.  Attempt to start a pod on a host that supports selinux (eg coreos).  To make it worse, use multiple containers (my test example has 3) in the pod that each mount the cephfs volume.

I know the above isn't very precise, I can try to put some specific commands together if there's difficulty reproducing.

**Environment**:
* OS (e.g. from /etc/os-release):
```
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1883.1.0
VERSION_ID=1883.1.0
BUILD_ID=2018-09-10-2343
PRETTY_NAME="Container Linux by CoreOS 1883.1.0 (Rhyolite)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
COREOS_BOARD="amd64-usr"
```

* Kernel (e.g. `uname -a`):
```
Linux localhost 4.14.69-coreos #1 SMP Mon Sep 10 22:45:12 UTC 2018 x86_64 Intel(R) Celeron(R) CPU N2820 @ 2.13GHz GenuineIntel GNU/Linux
```

* Cloud provider or hardware configuration:
baremetal

* Rook version (use `rook version` inside of a Rook Pod): v0.8.2

* Kubernetes version (use `kubectl version`):

Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.10", GitCommit:"098570796b32895c38a9a1c9286425fb1ececa18", GitTreeState:"clean", BuildDate:"2018-08-02T17:11:51Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/arm"}

* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
baremetal / kubeadm

* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/Rook/master/toolbox.html)):

It was HEALTH_OK during the above issues - right at this moment I'm mid upgrade to 0.8.3 so there's a lot of unrelated churn...

**Comment by anguslees** [comment]
After some brief [slack discussion](https://kubernetes.slack.com/archives/C09QZFCE5/p1538408298000100) with @gnufied, it was pointed out that the in-tree `cephfs` volume types does _not_ do selinux relabelling.  Following this line of thought, the below patch immediately fixes the above issue for me:

```patch
diff --git a/cmd/rookflex/cmd/init.go b/cmd/rookflex/cmd/init.go
index 81fcf443..abca32ff 100644
--- a/cmd/rookflex/cmd/init.go
+++ b/cmd/rookflex/cmd/init.go
@@ -36,7 +36,7 @@ func init() {
 }
 
 func initPlugin(cmd *cobra.Command, args []string) error {
-       fmt.Println(`{"status":"Success", "capabilities": {"attach": false}}`)
+       fmt.Println(`{"status":"Success", "capabilities": {"attach": false, "selinuxRelabel": false}}`)
        os.Exit(0)
        return nil
 }
```

I suspect a real patch would want to declare `selinuxRelabel=false` only for cephfs, not block volumes - assuming we want to continue to support selinux labels on block volumes.  Unfortunately I _think_ this means shipping separate plugin executables for the different ceph use cases, since capabilities are reported driver-wide (afaics).

There are still many open questions for me around when selinux relabelling is reasonable at all.  My limited selinux understanding suggests it might actively break ReadWriteMany volumes (last relabel wins), and as the above shows is only possible on "small" persistent volumes (where a recursive filesystem walk completes within a reasonable time).  I believe #sig-storage are also struggling to come up with a consistent strategy here.  (It seems a complete misfeature tbh...)

**Comment by gnufied** [comment]
in-tree has two separate drivers for cephfs and RBD. Can we not do something like that. I am not familiar with rook so apology if it is unworkable. cc @rootfs

**Comment by rootfs** [comment]
@anguslees  check this out https://tracker.ceph.com/issues/5486

cc @batrick for cephfs selinux latest status

**Comment by anguslees** [comment]
@rootfs: I think you might misunderstand the nature of my bug report.  Afaict cphfs _supports_ selinux labels just fine (setxattr).  The issue is that we don't _want_ docker to relabel cephfs volumes when they're used as kubernetes ReadWriteMany PVs.  As prior art, I offer the in-tree [nfs](https://github.com/kubernetes/kubernetes/blob/a4aab0e1e7fee44c712bd7031e118919389df2b3/pkg/volume/nfs/nfs.go#L227) and [cephfs](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/cephfs/cephfs.go#L209) volume types, which also avoid relabelling (see those links).

Indeed, if cephfs did _not_ support selinux labels then we wouldn't have to explicitly disable it like this ;)

**Comment by anguslees** [comment]
For anyone else stumbling across this: I've just spent the last week debugging what turned out to be the same issue all over again with cephfs (not rbd) CSI plugin  :(

Apparently you can't disable selinux relabelling in the CSI plugin (yet) - so your choices are to stick with the flexvolume, properly disable selinux in the kernel ('permissive' is not sufficient), or patch code.

**Comment by Madhu-1** [comment]
> For anyone else stumbling across this: I've just spent the last week debugging what turned out to be the same issue all over again with cephfs (not rbd) CSI plugin :(
> 
> Apparently you can't disable selinux relabelling in the CSI plugin (yet) - so your choices are to stick with the flexvolume, properly disable selinux in the kernel ('permissive' is not sufficient), or patch code.

@anguslees can you check  `container_use_cephfs` is set to false on the host machines? if yes we need to set it to on

```
getsebool -a | grep ceph
container_use_cephfs --> off
```
CSI driver won't have any role to play in SELinux as of now. is it possible to paste audit logs to see what's happening (or can you please provide steps to reproduce this issue).

**Comment by anguslees** [comment]
@Madhu-1 thanks for your response.

There is no such setting on my nodes.  (`getsebool -a | grep ceph` returns nothing)

rook v1.1.9 running ceph v14.2.6-20200115
on Flatcar Linux beta-channel nodes (currently 5.8.11 kernel with default flatcar selinux=permissive)
I have `ROOK_ENABLE_SELINUX_RELABELING=false` if that matters.

Symptoms: mons fell over and couldn't get back up.  One would go to 100% cpu for minutes, the in-progress election would timeout and restart, repeat.  Nothing in the logs, other than the election restarting frequently and the "slow ops" counter accumulating.   After days of much head scratching, and on a hunch from this earlier relabelling issue, I rebooted the single node that was attempting to mount my large cephfs/CSI PV with selinux disabled in the kernel, and the mons immediately recovered.

**Comment by anguslees** [comment]
> debugging what turned out to be the same issue all over again with cephfs

False alarm, sorry for the distraction and for not following up here earlier.  After summarising it above, it didn't sound right, so I went back and hunted again from the beginning.  Turned out my (latter) issue was https://tracker.ceph.com/issues/42830 and - as suggested in that bug - fixed by setting a much lower mon_sync_max_payload_size. Haven't had an issue since making that change.
